adr/ADR-41.md

```diff
@@ -0,0 +1,256 @@
+
+### Sorting
+
+These events form a Directed Acyclic Graph that you can sort using the Hop information. The Server that sends the message to another Route, Gateway or Leafnode will indicate it will publish to a number of `hops` and each server that receives the message will have the `Nats-Trace-Hop` header set in its `request`.  
+
+Given `Server` -> `Gateway` -> `Router` -> `Leafnode` -> `Client` the first remote side of the Gateway would have `1.1` and the Leafnode might have `1.1.1`. However if there were multiple subscribers and multiple hops you might end up with `1.2.3`. On the sending side the Egress events have a `hop` key that would be `1` in the first case etc.
+
+You would associate message either by using a unique `Nats-Trace-Dest` subject or by parsing the `traceparent` to get the trace and span IDs.
+
+### Trace Events
+
+The `events` list contains all the events that happened in a given server.  We see here that there are different types of event according to the table below:
+
+| Type | Server Data Type         | Description     |
+|------|--------------------------|-----------------|
+| `in` | `MsgTraceIngress`        | Ingress         |
+| `sm` | `MsgTraceSubjectMapping` | Subject Mapping |
+| `se` | `MsgTraceStreamExport`   | Stream Export   |
+| `si` | `MsgTraceServiceImport`  | Service Import  |
+| `js` | `MsgTraceJetStream`      | JetStream       |
+| `eg` | `MsgTraceEgress`         | Egress          |
+
+We also see a `kind` field, this holds the NATS Servers client Kind, at present these are the values:
+
+| Kind | Server Constant    | Description         |
+|------|--------------------|---------------------|
+| `0`  | `server.CLIENT`    | Client Connection   |
+| `1`  | `server.ROUTER`    | Router Connection   |
+| `2`  | `server.GATEWAY`   | Gateway Connection  |
+| `3`  | `server.SYSTEM`    | The NATS System     |
+| `4`  | `server.LEAF`      | Leafnode Connection |
+| `5`  | `server.JETSTREAM` | JetStream           |
+| `6`  | `server.ACCOUNT`   | Account             |
+
+You may not see all Kinds in traces but this is the complete current list.
+
+Using these Kinds and Types we can understand the JSON data above:
+
+ * `Ingress` from `Client` connection `cid:5390`
+ * `Egress` to `Client` connection `cid:5376`
+ * `Egress` to `Router` connection `rid:492` called `n1-lon` 
+
+This indicates a Client connected to this server (`n3-lon`) published the message it was received by a client connected to the same server and finally sent over a Route to a client on another server (`n1-lon`).
+
+### Specific Trace Types
+
+Most of the trace types are self explanatory but I'll call out a few specific here, the names match those in `msgtrace.go` in the Server source.
+
+#### MsgTraceEgress and MsgTraceIngress
+
+These messages indicates the message entering and leaving the server via a `kind` connection. It includes the `acc` it's being sent for, subscription (`sub`) and Queue Group (`queue`) on the Egress.
```

Comment:
We could add: `Note that for non CLIENT egresses, the subscription and queue group will be omitted. This is because NATS Servers optimize and send a single message across servers, based on a known interest, and let the remote servers match local subscribers. So it would be misleading to report the first match that caused a server to  forward a message across the route  as the egress' subscription.

Reply:
added thanks